libp2p/pubsub/pubsub.py

```diff
@@ -89,60 +56,54 @@ def __init__(self, host, router, my_id):
-        subs_map = {}
-        for topic in self.my_topics:
-            subs_map[topic] = True
-        sub_msg = MessageSub(
-            str(self.host.get_id()),\
-            str(self.host.get_id()), subs_map, generate_message_id()\
-        )
-        return sub_msg.to_str()
-
-    async def continously_read_stream(self, stream):
+        packet = rpc_pb2.RPC()
+        if self.my_topics:
+            for topic_id in self.my_topics:
+                packet.subscriptions.extend([rpc_pb2.RPC.SubOpts(
+                    subscribe=True, topicid=topic_id)])
+
+        return packet.SerializeToString()
+
+    async def continuously_read_stream(self, stream):
         """
         Read from input stream in an infinite loop. Process
-        messages from other nodes, which for now are considered MessageTalk
-        and MessageSub messages.
-        TODO: Handle RPC messages instead of my Aspyn's own custom message format
+        messages from other nodes
         :param stream: stream to continously read from
         """
+
+        # TODO check on types here
+        peer_id = str(stream.mplex_conn.peer_id)
+
         while True:
-            incoming = (await stream.read()).decode()
-            msg_comps = incoming.split('\n')
-            msg_type = msg_comps[0]
-
-            msg_sender = msg_comps[1]
-            # msg_origin = msg_comps[2]
-            msg_id = msg_comps[3]
-            print("HIT ME1")
-            if msg_id not in self.seen_messages:
-                print("HIT ME")
-                # Do stuff with incoming unseen message
-                should_publish = True
-                if msg_type == "subscription":
-                    self.handle_subscription(incoming)
-
-                    # We don't need to relay the subscription to our
-                    # peers because a given node only needs its peers
-                    # to know that it is subscribed to the topic (doesn't
-                    # need everyone to know)
-                    should_publish = False
-                elif msg_type == "talk":
-                    await self.handle_talk(incoming)
-
-                # Add message id to seen
-                self.seen_messages.append(msg_id)
-
-                # Publish message using router's publish
-                if should_publish:
-                    msg = create_message_talk(incoming)
-
-                    # Adjust raw_msg to that the message sender
-                    # is now our peer_id
-                    msg.from_id = str(self.host.get_id())
-
-                    await self.router.publish(msg_sender, msg.to_str())
+            incoming = (await stream.read())
+            rpc_incoming = rpc_pb2.RPC()
+            rpc_incoming.ParseFromString(incoming)
+
+            should_publish = False
+
+            if rpc_incoming.publish:
+                # deal with RPC.publish
+                for message in rpc_incoming.publish:
+                    if message.seqno not in self.seen_messages:
```

Comment:
Isn't it true that the [`seqno` could be duplicated across clients, but the corresponding messages could be different](https://github.com/libp2p/specs/tree/master/pubsub#the-message)? E.g. if A sends message "foo" with `seqno=1` and B sends "bar" with `seqno=1". In that case, would we fail to forward a message that we should have forwarded? I might be misunderstanding this code, if so feel free to clarify.

Reply:
this is correct, @stuckinaboot let's fix this one in #144 ?